File: android/src/main/kotlin/com/example/mongoatlasflutter/MongoAtlasClient.kt
package com.example.mongoatlasflutter

import com.google.android.gms.tasks.Task
import com.mongodb.client.MongoCollection
import com.mongodb.stitch.android.core.auth.StitchAuth
import com.mongodb.stitch.android.core.auth.StitchUser
import com.mongodb.stitch.android.services.mongodb.remote.RemoteFindIterable
import com.mongodb.stitch.android.services.mongodb.remote.RemoteMongoClient
import com.mongodb.stitch.android.services.mongodb.remote.RemoteMongoCollection
import com.mongodb.stitch.android.services.mongodb.remote.RemoteMongoDatabase
import com.mongodb.stitch.core.auth.providers.anonymous.AnonymousCredential
import com.mongodb.stitch.core.auth.providers.custom.CustomCredential
import com.mongodb.stitch.core.auth.providers.userpassword.UserPasswordCredential
import com.mongodb.stitch.core.services.mongodb.remote.RemoteDeleteResult
import com.mongodb.stitch.core.services.mongodb.remote.RemoteInsertOneResult
import io.flutter.plugin.common.StandardMessageCodec
import org.bson.BsonDocument
import org.bson.Document
import org.bson.conversions.Bson
import java.lang.Exception
import java.util.*
import kotlin.collections.HashMap



// Basic CRUD..

class MongoAtlasClient(
    private var client: RemoteMongoClient,
    private var auth: StitchAuth
) {


    /** ========================== Auth-related function  ========================= **/
    fun signInWithCustomJWT(jwtString: String): Task<StitchUser>? {
        return auth.loginWithCredential(CustomCredential(jwtString))
    }

    fun signInWithEmailPassword(username: String, password: String ): Task<StitchUser>? {
        return auth.loginWithCredential(UserPasswordCredential(username, password))
    }

    fun signInAnonymously(): Task<StitchUser>?
            = auth.loginWithCredential(AnonymousCredential())

    fun logout(): Task<Void> = auth.logout()

    fun isUserLoggedIn(): Boolean{
        return auth.user?.isLoggedIn ?: false
    }

    fun getUserId(): String?{
        return auth.user?.id
    }


    /** ========================== Database-related function  ========================= **/
    private fun getCollection(databaseName: String?, collectionName: String?): RemoteMongoCollection<Document>? {
        if(databaseName == null || collectionName == null)
            throw Exception()

        return client.getDatabase(databaseName).getCollection(collectionName)
    }

    fun insertDocument(databaseName: String?, collectionName: String?, data: HashMap<String, Any>?)
            : Task<RemoteInsertOneResult>? {
        val collection = getCollection(databaseName, collectionName)


        
        //Document.parse(json)
        val document = Document()

        if(data == null)
            return null

        for (item in data.entries){
            document[item.key] = item.value
        }

        return collection?.insertOne(document)
    }

//    fun insertDocuments(databaseName: String?, collectionName: String?, data: HashMap<String, Any>?)
//            : Task<RemoteInsertOneResult>? {
//        val collection = getCollection(databaseName, collectionName)
//
//        //Document.parse(json)
//        val document = Document()
//
//        if(data == null)
//            return null
//
//        for (item in data.entries){
//            document[item.key] = item.value
//        }
//
//        return collection?.insertOne(document)
//    }

    // TODO:  check this implementation
    fun deleteDocument(databaseName: String?, collectionName: String?, filterJson: String?)
            : Task<RemoteDeleteResult>? {
        val collection = getCollection(databaseName, collectionName)

        if (filterJson == null)
            return collection?.deleteOne(BsonDocument())

        val filter = BsonDocument.parse(filterJson)
        return collection?.deleteOne(filter)
    }

    // TODO:  check this implementation
    fun deleteDocuments(databaseName: String?, collectionName: String?, filterJson: String?)
            : Task<RemoteDeleteResult>? {
        val collection = getCollection(databaseName, collectionName)
        
        if (filterJson == null)
            return collection?.deleteMany(BsonDocument())

        val filter = BsonDocument.parse(filterJson)
        return collection?.deleteMany(filter)
    }

    /*******************************************************************************/

    fun findDocuments(databaseName: String?, collectionName: String?, filterJson: String?)
            : RemoteFindIterable<Document>? {
        val collection = getCollection(databaseName, collectionName)

        if (filterJson == null)
            return collection?.find()

        val filter = BsonDocument.parse(filterJson)
        return collection?.find(filter)
    }



    fun findDocument(databaseName: String?, collectionName: String?, filterJson: String?)
            : Task<Document>? {
        val collection = getCollection(databaseName, collectionName)
        
        if (filterJson == null)
            return collection?.findOne()

        val filter = BsonDocument.parse(filterJson)
        return collection?.findOne(filter)
    }


    fun countDocuments(databaseName: String?, collectionName: String?, filterJson: String?)
            : Task<Long>? {
        val collection = getCollection(databaseName, collectionName)

        if (filterJson == null)
            return collection?.count()
        
        val filter = BsonDocument.parse(filterJson)
        return collection?.count(filter)
    }

}

File: ios/Classes/MongoAtlasClient.swift
//
//  MongoAtlasClient.swift
//  mongoatlasflutter
//
//  Created by kfir Matit on 16/04/2020.
//

import Foundation
import MongoSwift
import StitchCore
import StitchRemoteMongoDBService

// cumbersome workaround solution
// TODO: convert any (possible) value into 'BSONValue'
class BsonExtractor {
    static func getValue(of: Any) -> BSONValue?{
        let value = of
        
        if let bsonValue = value as? String {
            return bsonValue
        }
        
        if let bsonValue = value as? Int {
            return bsonValue
        }
        
        
        if let bsonValue = value as? Double {
            return bsonValue
        }
        
        
        // TODO: check this conversion
        if let bsonValue = value as? Array<Any> {
            return bsonValue
        }
        
        return nil
    }
}


enum MyError : Error {
    case nilError
}


class MongoAtlasClient {
    var client: RemoteMongoClient
    var auth: StitchAuth
    
    init(client: RemoteMongoClient, auth: StitchAuth) {
        self.client = client
        self.auth = auth
    }
    
    
    func signInAnonymously(
        onCompleted: @escaping (Bool)->Void,
        onError: @escaping (String?)->Void
    ) {
        self.auth.login(withCredential: AnonymousCredential()) { authResult in
            switch authResult {
            case .success(let user):
                onCompleted(true)
                break
                
            case .failure(let error):
                onError("\(error)")
                break
            }
        }
    }
    
    func signInWithUsernamePassword(
        username: String,
        password: String,
        onCompleted: @escaping (Bool)->Void,
        onError: @escaping (String?)->Void
    ) {
        self.auth.login(
            withCredential: UserPasswordCredential(withUsername: username, withPassword: password)
        ) { authResult in
            switch authResult {
            case .success(let user):
                onCompleted(true)
                break
                
            case .failure(let error):
                onError("UsernamePassword Provider Login failed \(error)")
                break
            }
        }
    }
    
    func registerWithEmail(
        email: String,
        password: String,
        onCompleted: @escaping (Bool)->Void,
        onError: @escaping (String?)->Void
    ) {
        
        let emailPassClient = self.auth.providerClient(
            fromFactory: userPasswordClientFactory
        )
        
        emailPassClient.register(withEmail: email, withPassword: password) { result in
            switch result {
            case .success(let user):
                onCompleted(true)
                break
                
            case .failure(let error):
                onError("Error registering new user: \(error)")
                break
            }
        }
    }
    
    func logout(
        onCompleted: @escaping (Bool)->Void,
        onError: @escaping (String?)->Void
    ) {
        self.auth.logout { result in
            switch result {
            case .success(let user):
                onCompleted(true)
                break
                
            case .failure(let error):
                onError("Cannot logout user: \(error)")
                break
            }
        }
        
    }
    
    func getUserId() -> String? {
        return self.auth.currentUser?.id
    }
    
    /// ========================== Database related ========================================== ///
    private func getCollection(databaseName: String?, collectionName: String?) throws
        -> RemoteMongoCollection<Document>? {
        if(databaseName == nil || collectionName == nil) {
            throw MyError.nilError
        }
    
        return client.db(databaseName!).collection(collectionName!)
    }
    
    func insertDocument(
        databaseName: String?,
        collectionName: String?,
        data: Dictionary<String, Any>?,
        onCompleted: @escaping ()->Void,
        onError: @escaping (String?)->Void
   ) {
        do {
            let collection = try getCollection(databaseName: databaseName, collectionName: collectionName)
        
            var document = Document()
            for (key) in data!.keys{
                let value = data![key]
                
                if (value != nil){
                    document[key] = BsonExtractor.getValue(of: value!)
                }
            }
            
            
            collection?.insertOne(document) { result in
                switch result {
                case .success(let result):
                    print("Successfully inserted item with _id: \(result.insertedId))");
                    onCompleted()
                case .failure(let error):
                    onError("Failed to insert a document: \(error)")
                }
            }
        }
        catch {
            onError("Failed to insert a document")
        }
    }
    
    ///////////////////////////
 
    func deleteDocument(
        databaseName: String?,
        collectionName: String?,
        filterJson: String?,
        onCompleted: @escaping (Any?)->Void,
        onError: @escaping (String?)->Void
        ) {
        do {
            let collection = try getCollection(databaseName: databaseName, collectionName: collectionName)
            
            var document = Document()
            
            if (filterJson != nil){
                document = try Document.init(fromJSON: filterJson!)
            }
        
            
            collection?.deleteOne(document) { result in
                switch result {
                case .success(let result):
                    onCompleted(result)
                case .failure(let error):
                    onError("Failed to delete a document : \(error)")
                }
            }
        }
        catch {
            onError("Failed to delete a document")
        }
    }
    
    // CHECK
    func deleteDocuments(
        databaseName: String?,
        collectionName: String?,
        filterJson: String?,
        onCompleted: @escaping (Any?)->Void,
        onError: @escaping (String?)->Void
        ) {
        do {
            let collection = try getCollection(databaseName: databaseName, collectionName: collectionName)
            
            var document = Document()
            
            if (filterJson != nil){
                document = try Document.init(fromJSON: filterJson!)
            }
            
            collection?.deleteMany(document) { result in
                switch result {
                case .success(let result):
                    onCompleted(result)
                case .failure(let error):
                    onError("Failed to delete documents : \(error)")
                }
            }
        
        }
        catch {
            onError("Failed to delete documents")
        }
    }
  
  
    func findDocuments(
        databaseName: String?,
        collectionName: String?,
        filterJson: String?,
        onCompleted: @escaping (Any?)->Void,
        onError: @escaping (String?)->Void
        ) {
        do {
            let collection = try getCollection(databaseName: databaseName, collectionName: collectionName)
            
            var document = Document()
            
            if (filterJson != nil){
                document = try Document.init(fromJSON: filterJson!)
            }
            
            let task = collection?.find(document, options: nil)

            
            task!.toArray(){result in
                switch result {
                case .success(let results):
                    let queryResults =  results.map({ $0.extendedJSON })
                    onCompleted(queryResults)
                case .failure(let error):
                    onError("Failed to find documents: \(error)")
                }
            }
        }
        catch {
            onError("Failed to find documents")
        }
    }
    
 
    func findDocument(
        databaseName: String?,
        collectionName: String?,
        filterJson: String?,
        onCompleted: @escaping (Any?)->Void,
        onError: @escaping (String?)->Void
        ) {
        do {
            let collection = try getCollection(databaseName: databaseName, collectionName: collectionName)
            
            var document = Document()
            
            if (filterJson != nil){
                document = try Document.init(fromJSON: filterJson!)
            }
            
            collection?.findOne(document) { result in
                switch result {
                case .success(let result):
                    onCompleted(result?.extendedJSON)
                    break
                case .failure(let error):
                    onError("Failed to find item: \(error)")
                    break
                }
            }
        }
        catch {
            onError("Failed to find item")
        }
    }
    

    func countDocuments(
        databaseName: String?,
        collectionName: String?,
        filterJson: String?,
        onCompleted: @escaping (Any)->Void,
        onError: @escaping (String?)->Void
        ) {
        do {
            let collection = try getCollection(databaseName: databaseName, collectionName: collectionName)
            
            var document = Document()
        
            if (filterJson != nil){
                document = try Document.init(fromJSON: filterJson!)
            }

            
            collection?.count(document) { result in
                switch result {
                case .success(let result):
                    onCompleted(result)
                case .failure(let error):
                    onError("Failed to count collection: \(error)")
                }
            }
        
        }
        catch {
            onError("Failed to count collection")
        }
    }
}
